mitmproxy/controller.py

```diff
@@ -147,23 +167,63 @@ def __call__(self, msg=False):
 NO_REPLY = object()
 
 
-class Reply(object):
+def handler(f):
+    @functools.wraps(f)
+    def wrapper(*args, **kwargs):
+        if len(args) == 1:
+            message = args[0]
+        elif len(args) == 2:
+            message = args[1]
+        else:
+            raise ControlError("Handler takes one argument: a message")
+
+        if not hasattr(message, "reply"):
+            raise ControlError("Message %s has no reply attribute"%message)
+
+        handling = False
+        # We're the first handler - ack responsibility is ours
+        if not message.reply.handled:
+            handling = True
+            message.reply.handled = True
+
+        ret = f(*args, **kwargs)
 
+        if handling and not message.reply.acked and not message.reply.taken:
+            message.reply()
+        return ret
+    wrapper.func_dict["__handler"] = True
```

Comment:
Using func_dict is mildly more explicit about what we're doing. I don't have strong feelings here.